### ISSUE TITLE
Dataset Files DataTable - Paginator, Selection, Download, Counter

### ISSUE THREAD

**Issue by mheppler** [IQSS/dataverse#4656]
Based on feedback from @scolapasta @TaniaSchlatter and @djbrooke on the issue _Spike: Files DataTable, Selection, Paginator vs Scroll #4437_, creating a new issue for development.

Revise the dataset files table to improve UX of the table with improvements to paginator, selection, counter, download and potentially sort features of the dataset pg files table. Bug and feature request issues have been consolidated and included in _Dataset - File UI Improvements #3404_.

If there is anything we learn about performance of this files table, we should add it to this issue, _Performance degradation investigation #3631_.

Defined scope:
- Swap out dataTable liveScroll attribute and instead use the Paginator (out of box paginator will need CSS hacking to look more Bootstrapy)
- Dynamic file counter that shows how many files are being displayed in the table, in addition to how many files total, and how many files are selected
- Text links for selecting all and clearing selections displayed along with the dynamic count of how many files are currently selected
- Review how the current Download button in the dataTable header works with relation to the "select all" hack (#4051) added to the checkbox in the dataTable header

![files - dataset - depositor jpeg](https://user-images.githubusercontent.com/687227/39823277-be3ac73c-537a-11e8-90ba-0351b9f00c48.png)

**Comment by mheppler** [comment]
Regarding performance, there can be front-end refactoring improvements to optimize the XHTML, towards small gains in render and load time.

For example, in `filesFragment.xhtml` include file for the dataTable, there are 16 references to the `DatasetPage.editMode` render logic, but in `dataset.xhtml` there is render logic for `empty DatasetPage.editMode` that does not display the tab containing the datatable in the include.

```
<p:tab id="dataFilesTab" title="#{bundle.files}" rendered="#{(!DatasetPage.workingVersion.deaccessioned or 
                                                              (DatasetPage.workingVersion.deaccessioned and DatasetPage.canUpdateDataset()))  and
                                                              (empty DatasetPage.editMode)}">
                                <ui:include src="filesFragment.xhtml">
                                    <ui:param name="fileDownloadHelper" value="#{DatasetPage.fileDownloadHelper}"/>
                                </ui:include>
</p:tab>
```

**Comment by djbrooke** [comment]
We had a good discussion about this in Sprint Planning today but there was not consensus on the scope and the approach to take. If anyone from the team is interested in championing this and bringing it back for a future Backlog Grooming or Sprint Planning meeting, I'm happy to facilitate a meeting or other discussions.

**Comment by adam3smith** [comment]
Thanks for working on this again! I love the "Download all files" buttons, but for the same reasons brought up on the [recent community group thread](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/dataverse-community/pBDLtYQaY1A/VNGddwYzBgAJ), it'd be very helpful to also have an actually working "select all" button. The fact that the select all button works only for visible files is at best inkonvenient (that's for us at QDR as experienced curators) and at worst dangerous, as per the use case reported on the list where a self-depositing user doesn't properly restrict access to all files.

From the thread:

> I've just discovered this issue and it is a HUGE issue. If you're restricting files and click the select all box it doesn't restrict all the files in the dataset unless you've scrolled down first. This can cause major problems. This is not clear at all.

**Comment by TaniaSchlatter** [comment]
We reviewed and defined behavior for selecting/checkboxes:

- checkboxes remain selected across pages in the dataset

- download button text is dynamic and changes to indicate the number of items (selected) that will be downloaded

- selecting the checkbox at the top left of the table selects all visible items and provides feedback indicating the number of items selected, the total number of items in the dataset, ability to select all items in the dataset, ability to deselect (select none) all items in the dataset.

**Comment by TaniaSchlatter** [comment]
See the 4th bullet above, @pdurbin, for how we decided to provide "select all" functionality at this time.

**Comment by mheppler** [comment]
Putting this on ice so that we can get _UI - Primefaces 6.2 Upgrade #4220_ merged to take advantage of the [paginator features](https://www.primefaces.org/showcase/ui/data/datatable/paginator.xhtml) that were (maybe?) introduced in 6.1.

The `paginatorTopLeft` facet in the showcase example allows for buttons to easily be added and positioned in the dataTable header.

```
        <f:facet name="paginatorTopLeft">
            <p:commandButton type="button" icon="fa fa-sync" />
        </f:facet>
```

Here is a screenshot of the current progress of the UI. It is pretty close to ready. Other than getting the "Upload Files" and the dynamic counter positioned correctly compared to the mockups, there isn't a whole lot left. There were some bugs related to the dynamic counter and the select-all checkbox, which I had mentioned to @sekmiller as a heads up for future swarming.

![screen shot 2018-06-20 at 3 58 04 pm](https://user-images.githubusercontent.com/687227/41681519-d4c54b78-74a2-11e8-9ac7-634dbe6de20e.png)

**Comment by mheppler** [comment]
Found an example of the dataTable dynamic counter work that was done for the [File Permissions pg](https://github.com/IQSS/dataverse/blob/develop/src/main/webapp/permissions-manage-files.xhtml) as part of #2716.

```
                            <p:outputPanel id="selectionRemoveCounter">
                                <p>
                                    <h:outputFormat styleClass="highlightBold" value="#{bundle['dataverse.permissionsFiles.files.selected']}">
                                        <f:param value="#{manageFilePermissionsPage.selectedRoleAssignmentRows.size()}"/>
                                        <f:param value="#{manageFilePermissionsPage.roleAssignments.size()} "/>
                                        <f:param value="#{!empty manageFilePermissionsPage.selectedFile ? bundle['dataverse.permissionsFiles.files.roleAssignees'] : bundle['dataverse.permissionsFiles.files']}"/>
                                    </h:outputFormat>
                                </p>
                            </p:outputPanel>
                            
                            <p:dataTable id="assignedRoles" var="roleAssignment" value="#{manageFilePermissionsPage.roleAssignments}"
                                         selection="#{manageFilePermissionsPage.selectedRoleAssignmentRows}" rowKey="#{roleAssignment.id}" 
                                         sortBy="#{!empty manageFilePermissionsPage.selectedFile ? roleAssignment.assigneeDisplayInfo.title : roleAssignment.definitionPoint.displayName}"
                                         scrollable="true" scrollHeight="300" disabledSelection="#{!roleAssignment.definitionPoint.instanceofDataFile}">
                                
                                <p:ajax event="rowSelectCheckbox"  update=":rolesPermissionsForm:selectionRemoveCounter" />
                                <p:ajax event="rowUnselectCheckbox"  update=":rolesPermissionsForm:selectionRemoveCounter" />
                                <p:ajax event="toggleSelect"  update=":rolesPermissionsForm:selectionRemoveCounter" />
                                
                                <p:column selectionMode="multiple" style="width:2%;text-align:center"/>
                                <p:column width="98%" headerText="#{bundle['dataverse.permissionsFiles.usersOrGroups.tabHeader.userOrGroup']}" rendered="#{!empty manageFilePermissionsPage.selectedFile}">
                                    <h:outputText value="#{roleAssignment.assigneeDisplayInfo.title}"/> <h:outputText value=" (#{roleAssignment.assigneeDisplayInfo.affiliation})" rendered="#{!empty roleAssignment.assigneeDisplayInfo.affiliation}"/>
                                </p:column>
                                <p:column width="98%" headerText="#{bundle['dataverse.permissionsFiles.files.tabHeader.fileName']}" rendered="#{!empty manageFilePermissionsPage.selectedRoleAssignee}">
                                    <h:outputText value="#{roleAssignment.definitionPoint.displayName}"/>
                                </p:column>
                            </p:dataTable>
```
The bugs with the dynamic counter the File Permissions pg seem similar to the bugs with the dataset pg, so hopefully this solution also applies here.

**Comment by mheppler** [comment]
Also, found a minor bug locally and on [demo (4.9)](https://demo.dataverse.org/dataset.xhtml?persistentId=doi:10.5072/FK2/WYQYCO) related to the files dataTable and selection. The file thumbnail preview is broken in the files dataTable on the dataset pg when you select files, the page refreshes, and you then hover over the thumbnail to show the preview popover.

**Comment by mheppler** [comment]
Reminder, check `structure.css` for `ui-datatable-scrollable` class name references and revise/remove those as necessary.

```
div[id$="filesTable"].ui-datatable.ui-datatable-scrollable {padding-right:12px;}

#file-replace-original-block div[id$="filesTable"].col-sm-10 {border:1px solid #ddd;padding:0;}
#file-replace-original-block div[id$="filesTable"] .col-file-thumb {width:74px !important;padding:4px 10px;}
#file-replace-original-block div[id$="filesTable"] .col-file-metadata {padding:4px 10px;}

/* HACK */
div[id$="filesTable"] div.ui-datatable-scrollable-body > table, div[id$="filesTable"] div.ui-datatable-tablewrapper > table {margin-bottom:140px;}
```

**Comment by pdurbin** [comment]
@mheppler and I just discussed this issue. To start the plan is to merge #4780 (4220-primefaces-6-2) and #4806 (4219-bootstrap-upgrade) into the 4656-filetable-paginator branch and then put the code up on https://dev1.dataverse.org so that we call all look at the GUI and decide what the remaining TODO items are for this issue. At a high level I think of this issue as providing relief from the "live scroll" we added in c9b0e1c (v4.2.1 - https://github.com/IQSS/dataverse/releases/tag/v4.2.1 ) which has been a bit problematic in a variety of ways.

**Comment by mheppler** [comment]
Somewhat related to selecting files + download, _Restricted Files Download - Fix Height for Warning Popup with Large File List #4472_.

**Comment by pdurbin** [comment]
I just did all the merging described above and below is a screenshot as of 5c3b1fc with Primefaces 6.2 and Bootstrap 3.3.7 from https://dev1.dataverse.org/dataset.xhtml?persistentId=doi:10.5072/FK2/WGWEOM

![42_files_-_root_-_2018-07-11_13 02 59](https://user-images.githubusercontent.com/21006/42588197-1ec515fa-850b-11e8-81ea-92040d443aef.png)

Obviously, we need to remove "TEST", but what else should go on the todo list before we can move this issue to code review?

**Comment by pdurbin** [comment]
@mheppler and I just came up with this TODO list:

- [ ] Remove one of the duplicate "selection messages" (i.e. "0 files are currently selected. Select all 42 files in this dataset. Clear selection.").
- [ ] Remove "TEST" and move dynamic counter to be left-aligned.
- [ ] Right align "Upload Files" button.
- [ ] Styling of the paginator has been lost. Make more Bootstrappy.
- [ ] Make "Clear selection" link work (actually do something).
- [ ] Make "Select all NN files" link work (actually do something).
- [ ] Select all checkbox (in table header) bugs: Clicking checkbox doesn't change selection message.

**Comment by pdurbin** [comment]
@mheppler and I just discussed the state of the code as of 2fcc12f and poked around where it is deployed at https://dev1.dataverse.org

We agreed that he's going to go ahead and make commits to fix the left-align, right-align, and styling items in the checklist above so I'll add him to this issue.

I'll leave myself on the issue for now. I will need to pick @sekmiller 's brain because Mike wants us to undo the hacks added in pull request #4363 for the "select all" checkbox. He and I also noticed a bug with file selection counts when you start using the "select all" checkbox.

**Comment by pdurbin** [comment]
I just showed @sekmiller the screen recording above and he said he'd take a look. The branch is "4656-filetable-paginator". We also tried with no files selected and only clicking the "select all" checkbox results in "0 file(s) are currently selected".

**Comment by pdurbin** [comment]
@sekmiller fixed the bug above in 45981b3 and he and I have resolved to stop calling that checkbox the "select all" checkbox because it's misleading. When a dataset has 42 files "select all" implied that all files will be selected. That's not what the checkbox does. It selects all the files that are visible on the current page. So if you want all the files on page 2 and 4, you navigate to page 2, check the box, navigate to page 3 (the box will no longer be checked), navigate to page 4 (still unchecked), and the check the box again, bringing your total count of selected files to 20 (using the default of 10 files per page). I guess we could call it the "select all files on page" checkbox but that's a mouthful. Thankfully, it behaves more or less the same as Gmail, so that's a good mental model for it.

Related to this, @sekmiller and I just double-checked with @djbrooke that for this issue we are definitely reverting the hacks in pull request #4363 and so undoing the change to the behavior of that checkbox that we implemented in #4051. Since @sekmiller made that pull request he is in a much better place than I am to back it out and test that nothing was broken in the process and he's happy to do it, so I'm removing myself from this issue. I'll add an item above to revert the pull request in the TODO list above. @mheppler is still working on the UI related fixes.

**Comment by pdurbin** [comment]
I put the latest code (63efaf5) on https://dev1.dataverse.org and @TaniaSchlatter @jggautier @djbrooke @scolapasta @mheppler @dlmurphy and I just talked through the current state during a design meeting. Here's a screenshot:

<img width="790" alt="screen shot 2018-07-17 at 1 02 29 pm" src="https://user-images.githubusercontent.com/21006/42833305-b7f1e2ca-89c1-11e8-9757-c9bc8227f451.png">

@mheppler talked through the need to have more information above the table and @scolapasta suggested looking at https://stackoverflow.com/questions/17487670/update-component-at-change-of-page-on-primefaces-datagrid-with-pagination

**Comment by mheppler** [comment]
In #2506, there was a commit 5367b8ca0a3164b1bd8d3700e77dae72c42c90c4 to disable a sort button on the dataset pg files table, because it was too buggy to ship. The button code and render logic was left in the XHTML however. I am removing it in this branch, and we can revisit this feature as part of #2506.

**Comment by pdurbin** [comment]
I spoke with @mheppler this morning and it sounds like we're going to try https://stackoverflow.com/questions/17487670/update-component-at-change-of-page-on-primefaces-datagrid-with-pagination so I guess I'll give it a shot.

**Comment by mheppler** [comment]
The two checklist items in the comment from Phil still require attention. In my latest commit, I added a `TO-DO` list into `filesFragment.xhtml` above the static counter I returned to the top of the dataTable, so that it may be dynamisized(?).

- [ ] Convert static "1 to 10 of..." to dynamic counter wired to dataTable paginator with p:ajax
- [ ] Use h:outputFormat and f:param as we see here, with {0} and {1} in the bundle message rather than multiple bundle strings currently used in currentPageReportTemplate
- [ ] Conditional display dynamic "# to # of..." if the paginator is required (e.g. "1 File", "10 Files", "1 to 10 of 11 Files")
- [ ] Remove {CurrentPageReport} from paginatorTemplate attribute, as well as currentPageReportTemplate attribute
- [ ] Sanity check configurable Compute UI elements above the dataTable to make sure we didn't break nothing
- ?Fix misleading "# Files Selected on Upload Files/Edit Files pg?
- ?Copy over new selection message to Upload Files/Edit Files pg?

That should be every thing. Will meet with @pdurbin in person to make sure he's on board.

**Comment by pdurbin** [comment]
As I mentioned at standup, I'm blocked trying to implement the StackOverflow suggestion above. In 8e1796f I pushed my progress with FIXMEs but I need help from someone who knows and loves JSF.

**Comment by pdurbin** [comment]
@sekmiller is working on this: Revert the hacks in pull request #4363 to undo the change to the behavior of the "select all files on page" checkbox that we implemented in #4051.

He also tried to implement the StackOverflow suggestion above and couldn't get it working. (I spent about three hours trying to get it to work). Since we're both blocked, I'm adding @scolapasta to unblock us. Either we get it working or we need a new plan/design.

**Comment by sekmiller** [comment]
I have been able to get the update of the files displayed to work when the page is changed (1-10 changed to 11-20 when you go to page 2). Still working on the update when change is from 10 files per page to 25 or 50.

**Comment by pdurbin** [comment]
@sekmiller great news! You don't seem to be blocked to I narrowed the assignees for this issue to be just you. If there's an opportunity to swarm on this, please let me know!

**Comment by pdurbin** [comment]
I just deployed 53415d6 to https://dev1.dataverse.org if anyone wants to take a look. Thanks, @sekmiller !!! You are my JSF hero.

There's still a little cleanup Mike wanted so I'm going to jump on the branch and see what I can do.

**Comment by pdurbin** [comment]
I just updated the to do lists above and made pull request #4881 and will move this to code review.

One of the items was to sanity check how the compute button looks so here's a screenshot:

![screen shot 2018-07-23 at 1 00 07 pm](https://user-images.githubusercontent.com/21006/43091819-2cc31c20-8e79-11e8-9653-103f3c11df69.png)

**Comment by kcondon** [comment]
Generally works well. Found a few file count display issues:

- [ ] Changing the display size can corrupt the file count
1,901 to 1,000 of 1,000 Files

To reproduce:
-upload 1000 files (probably would happen with less but >2x50
-Click on button to go to end of pages.
-Change files per page to 25, then to 50. Notice count displayed is recalculated incorrectly.

- [ ] Uploading files on dataset creation displays an unusual file count on upload page:
51 to {1} of {2} {2}

- [ ] Upload files after dataset creation says files selected on upload page, rather than just files as in previous version:
10 Files Selected
